src/go/mongolib/stats/stats.go

```diff
@@ -0,0 +1,268 @@
+package stats
+
+import (
+	"crypto/md5"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/montanaflynn/stats"
+	"github.com/percona/percona-toolkit/src/go/mongolib/fingerprinter"
+	"github.com/percona/percona-toolkit/src/go/mongolib/proto"
+	"github.com/percona/percona-toolkit/src/go/mongolib/util"
+)
+
+// New creates new instance of stats with given fingerprinter
+func New(fingerprinter fingerprinter.Fingerprinter) *Stats {
+	s := &Stats{
+		fingerprinter: fingerprinter,
+	}
+
+	s.Reset()
+	return s
+}
+
+// Stats is a collection of MongoDB statistics
+type Stats struct {
+	// dependencies
+	fingerprinter fingerprinter.Fingerprinter
+
+	// internal
+	queryInfoAndCounters map[GroupKey]*QueryInfoAndCounters
+}
+
+// Reset clears the collection of statistics
+func (s *Stats) Reset() {
+	s.queryInfoAndCounters = make(map[GroupKey]*QueryInfoAndCounters)
+}
+
+// Add adds proto.SystemProfile to the collection of statistics
+func (s *Stats) Add(doc proto.SystemProfile) error {
+	fp, err := s.fingerprinter.Fingerprint(doc.Query)
+	if err != nil {
+		return fmt.Errorf("cannot get fingerprint: %s", err)
+	}
+	var qiac *QueryInfoAndCounters
+	var ok bool
+
+	key := GroupKey{
+		Operation:   doc.Op,
+		Fingerprint: fp,
+		Namespace:   doc.Ns,
+	}
+	if qiac, ok = s.queryInfoAndCounters[key]; !ok {
+		realQuery, _ := util.GetQueryField(doc.Query)
+		qiac = &QueryInfoAndCounters{
+			ID:          fmt.Sprintf("%x", md5.Sum([]byte(fmt.Sprintf("%s", key)))),
+			Operation:   doc.Op,
+			Fingerprint: fp,
+			Namespace:   doc.Ns,
+			TableScan:   false,
+			Query:       realQuery,
+		}
+		s.queryInfoAndCounters[key] = qiac
+	}
+	qiac.Count++
+	qiac.NScanned = append(qiac.NScanned, float64(doc.DocsExamined))
+	qiac.NReturned = append(qiac.NReturned, float64(doc.Nreturned))
+	qiac.QueryTime = append(qiac.QueryTime, float64(doc.Millis))
+	qiac.ResponseLength = append(qiac.ResponseLength, float64(doc.ResponseLength))
+	var zeroTime time.Time
+	if qiac.FirstSeen == zeroTime || qiac.FirstSeen.After(doc.Ts) {
+		qiac.FirstSeen = doc.Ts
+	}
+	if qiac.LastSeen == zeroTime || qiac.LastSeen.Before(doc.Ts) {
+		qiac.LastSeen = doc.Ts
+	}
+
+	return nil
+}
+
+// Queries returns all collected statistics
+func (s *Stats) Queries() Queries {
+	return mapToArray(s.queryInfoAndCounters)
+}
+
+// Queries is a slice of MongoDB statistics
+type Queries []QueryInfoAndCounters
+
+// CalcQueriesStats calculates QueryStats for given uptime
+func (q Queries) CalcQueriesStats(uptime int64) []QueryStats {
```

Comment:
It's ok, I'm using this one, I'm not using the other one but still I think this is quite correct place for this functions. 👍